README.md
# saucepan

Data structures and utilities for dealing with source spans.


![saucepan](saucepan.png)



### Requirements

 * Integration with error reporting (`codespan_reporting` or equivalent)
 * lightweight span
 * span can be input type (i.e. gives access to slice)
 * span can also give (row, col) info
 * unambiguous multi-file support
 
Would also be nice to have generic input type, but just `&str` is ok.

### Existing libraries

1. [`nom_locate`](https://crates.io/crates/`nom_locate`)
    1. Excellent integration with `nom`.
    2. generic input type
    3. minimal location types
    4. Includes `Source`/`Sources` types
    5. _No support for reporting_
    6. _lots of unsafe_ (a requirement of its design decisions)
2. [`codespan`](https://crates.io/crates/codespan)
    1. Native integration with `codespan_reporting`
    2. lightweight
    3. _no buffer/slice, thus cannot be used as input_
    4. Otherwise full-featured.
3. [`source-span`](https://crates.io/crates/source-span)
    1. Very nice, full-featured. Includes:...
    2. error reporting
    3. Text buffer
    4. Location types
    5. _allocating_
4. [`rls-span`](https://crates.io/crates/rls-span)
    1. Excellent for its purpose
    2. _Way_ more abstractions than necessary (`ZeroIndexed`, `OneIndexed` subtraits of `Indexed`)
    3. _Poorly documented_
    4. _No slice/buffer — location only_
    5. _No error reporting_



Saucepan _was_ a Franken-crate, a mash-up of [codespan](https://crates.io/crates/codespan) and
[`nom_locate`](https://crates.io/crates/`nom_locate`). After some attempts and thought, it became clear
that trying to marry the existing libraries just won't work. They are designed for different tasks,
and the design decisions that make them good at those tasks make them bad for each other. It was a
toxic relationship, so I broke it up.


## Is saucepan right for you?

### Saucepan's use case

Saucepan's use case is a situation in which a single type needs to serve as both an input slice type
compatible with `nom` and a span type (potentially compatible with `codespan_reporting` or
equivalent).

### when to use something else

If you only need `nom_locate` or codespan, or if your application can just as easily use both
`nom_locate` and `codespan` to satisfy its needs you probably shouldn't use this crate.


This crate should not be used in production at this time. Potential contributors should look at the
TODO.md file and the `// todo` comments if they want guidance on what to improve.

## ...why?

The purpose of this crate is to have a single *thing* that can take the place of both
 `nom_locate::LocatedSpan`
 and `codespan::Span`. That thing is called Span in this crate. A span can
 
 * be used as an input/output type for `nom` parser combinators
 * be queried for the byte offset, line (row) number, and column number of the text it represents
 * provide a `&str` of the text it represents
 * keep track of the file handle of the original source file the text is from
 * support a mechanism to retrieve the name of the source file
 * be lightweight, implement `Copy`
 * intgrate with `codespan_reporting` or equivalent

The `Span` struct itself implements copy.

## Authors and License

Copyright (C) Robert Jacobson

Released under the
Apache License Version 2.0. See [LICENSE](LICENSE) for details.

This work may contain code derived from `codespan`, which is published under the Apache 2.0 license
(see the file LICENSE), and `nom_locate`, which is published under the MIT license (see the file
LICENSE_MIT).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

* extracted from the original `codespan` repo
* `File` is now `pub`lic
* Renamed `File` —> `Source` and `Files` —> `Sources` and related names, as source code need not come from a file.
* `Span` and `Source` (prev. `File`) now optionally know their `SourceID`.
* `SourceID` (prev. `FileId`) is generic in `Sources`.



## [0.9.5] - 2020-06-24

### Fixed

-   Building the crate with the `reporting` feature disabled.

## [0.9.4] - 2020-05-18
## [0.9.3] - 2020-04-29
## [0.9.2] - 2020-03-29

## [0.9.1] - 2020-03-23

### Added

-   `From<Span>` is now implemented for `Range<usize>` and `Range<RawIndex>`.
-   More `From<_>` conversions are now implemented for index and offset types.

## [0.9.0] - 2020-03-11

### Added

-   `codespan` now depends on `codespan_reporting`. This can be disabled via the `reporting` feature.
-   `codespan::Files` now implements `codespan_reporting::files::Files`

## [0.8.0] - 2020-02-24
## [0.7.0] - 2020-01-06
## [0.6.0] - 2019-12-18
## [0.5.0] - 2019-10-02
## [0.4.1] - 2019-08-25
## [0.4.0] - 2019-08-22
## [0.3.0] - 2019-05-01
## [0.2.1] - 2019-02-26
## [0.2.0] - 2018-10-11

[Unreleased]: https://github.com/brendanzab/codespan/compare/v0.9.5...HEAD
[0.9.4]: https://github.com/brendanzab/codespan/compare/v0.9.4...v0.9.5
[0.9.4]: https://github.com/brendanzab/codespan/compare/v0.9.3...v0.9.4
[0.9.3]: https://github.com/brendanzab/codespan/compare/v0.9.2...v0.9.3
[0.9.2]: https://github.com/brendanzab/codespan/compare/v0.9.1...v0.9.2
[0.9.1]: https://github.com/brendanzab/codespan/compare/v0.9.0...v0.9.1
[0.9.0]: https://github.com/brendanzab/codespan/compare/v0.8.0...v0.9.0
[0.8.0]: https://github.com/brendanzab/codespan/compare/v0.7.0...v0.8.0
[0.7.0]: https://github.com/brendanzab/codespan/compare/v0.6.0...v0.7.0
[0.6.0]: https://github.com/brendanzab/codespan/compare/v0.5.0...v0.6.0
[0.5.0]: https://github.com/brendanzab/codespan/compare/v0.4.1...v0.5.0
[0.4.1]: https://github.com/brendanzab/codespan/compare/v0.4.0...v0.4.1
[0.4.0]: https://github.com/brendanzab/codespan/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/brendanzab/codespan/compare/v0.2.1...v0.3.0
[0.2.1]: https://github.com/brendanzab/codespan/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/brendanzab/codespan/releases/tag/v0.2.0


src/shims.rs
/*!
Shims for the case that the feature `nom-parsing` is disabled. This module is conditionally
included in `lib.rs`.

This code is taken from `nom::traits`.
*/

use std::ops::{
  Range,
  RangeFrom,
  RangeTo,
  RangeFull,
};



/// Helper trait for types that can be viewed as a byte slice
pub trait AsBytes {
  /// casts the input type to a byte slice
  fn as_bytes(&self) -> &[u8];
}

impl<'a> AsBytes for &'a str {
  #[inline(always)]
  fn as_bytes(&self) -> &[u8] {
    <str as AsBytes>::as_bytes(self)
  }
}

impl AsBytes for str {
  #[inline(always)]
  fn as_bytes(&self) -> &[u8] {
    self.as_ref()
  }
}

impl<'a> AsBytes for &'a [u8] {
  #[inline(always)]
  fn as_bytes(&self) -> &[u8] {
    *self
  }
}

impl AsBytes for [u8] {
  #[inline(always)]
  fn as_bytes(&self) -> &[u8] {
    self
  }
}

macro_rules! as_bytes_array_impls {
  ($($N:expr)+) => {
    $(
      impl<'a> AsBytes for &'a [u8; $N] {
        #[inline(always)]
        fn as_bytes(&self) -> &[u8] {
          *self
        }
      }

      impl AsBytes for [u8; $N] {
        #[inline(always)]
        fn as_bytes(&self) -> &[u8] {
          self
        }
      }
    )+
  };
}

as_bytes_array_impls! {
     0  1  2  3  4  5  6  7  8  9
    10 11 12 13 14 15 16 17 18 19
    20 21 22 23 24 25 26 27 28 29
    30 31 32
}


/// slicing operations using ranges
///
/// this trait is loosely based on
/// `Index`, but can actually return
/// something else than a `&[T]` or `&str`
pub trait Slice<R> {
  /// slices self according to the range argument
  fn slice(&self, range: R) -> Self;
}

macro_rules! impl_fn_slice {
  ( $ty:ty ) => {
    fn slice(&self, range: $ty) -> Self {
      &self[range]
    }
  };
}

macro_rules! slice_range_impl {
  ( [ $for_type:ident ], $ty:ty ) => {
    impl<'a, $for_type> Slice<$ty> for &'a [$for_type] {
      impl_fn_slice!($ty);
    }
  };
  ( $for_type:ty, $ty:ty ) => {
    impl<'a> Slice<$ty> for &'a $for_type {
      impl_fn_slice!($ty);
    }
  };
}

macro_rules! slice_ranges_impl {
  ( [ $for_type:ident ] ) => {
    slice_range_impl! {[$for_type], Range<usize>}
    slice_range_impl! {[$for_type], RangeTo<usize>}
    slice_range_impl! {[$for_type], RangeFrom<usize>}
    slice_range_impl! {[$for_type], RangeFull}
  };
  ( $for_type:ty ) => {
    slice_range_impl! {$for_type, Range<usize>}
    slice_range_impl! {$for_type, RangeTo<usize>}
    slice_range_impl! {$for_type, RangeFrom<usize>}
    slice_range_impl! {$for_type, RangeFull}
  };
}

slice_ranges_impl! {str}
slice_ranges_impl! {[T]}
